libbpfgo/libbpfgo.go

```diff
@@ -598,14 +621,79 @@ func doAttachKprobeLegacy(prog *BPFProg, kp string, isKretprobe bool) (*BPFLink,
 var eventChannels = make(map[uintptr]chan []byte)
 var lostChannels = make(map[uintptr]chan uint64)
 
+func (m *Module) InitRingBuf(mapName string, eventsChan chan []byte) (*RingBuffer, error) {
+	bpfMap, err := m.GetMap(mapName)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := uintptr(bpfMap.fd)
+	if eventsChan == nil {
+		return nil, fmt.Errorf("events channel can not be nil")
+	}
+	eventChannels[ctx] = eventsChan
+
+	rb := C.init_ring_buf(bpfMap.fd)
+	if rb == nil {
+		return nil, fmt.Errorf("")
+	}
+
+	ringBuf := &RingBuffer{
+		rb:     rb,
+		bpfMap: bpfMap,
+		stop:   make(chan bool),
+	}
+	m.ringBufs = append(m.ringBufs, ringBuf)
+	return ringBuf, nil
+}
+
+func (rb *RingBuffer) Start() {
+	go rb.poll()
+}
+
+func (rb *RingBuffer) Stop() {
+	rb.stop <- true
+}
+
+func (rb *RingBuffer) Close() {
+	C.ring_buffer__free(rb.rb)
+}
+
+func (rb *RingBuffer) poll() error {
+
+	pollChannel := make(chan C.int)
+
+	go func() {
+		for {
+			err := C.ring_buffer__poll(rb.rb, -1)
+			pollChannel  <- err
+		}
+	}()
+
+	for {
+		select {
+		case <-rb.stop:
+			return nil
+		case err := <-pollChannel:
```

Comment:
Why not using timeout instead? (like we do for the perf buffer)
`err := C.ring_buffer__poll(rb.rb, 300)`

Reply:
I'm trying to figure out what is the exact unexpected behavior that's causing `ring_buffer__poll` to block regardless of the specified timeout. The test program I wrote exposed a block where the write to `rb.stop` in ringbuffer's `Stop` method to block indefinitely. Perhaps a channel was closed somewhere that the poll function was caught trying to write to. Perhaps a reused map file descriptor that's not being properly cleaned up. 

Pushing up that test program now so you can take a look.

Reply:
didn't really investigate this but just noting that it's impossible to write to a closed channel, more likely it's a double write to an unbuffered channel